File: python/LinkedList/LinkedList.py
class Node:
    def __init__(self, value):
        #super().__init__()
        self.value = value
        self.next = None
    
    
class SingleLinkedList:
    def __init__(self):
        #super().__init__()
        self.head = None
        self.size = 0
    
    def __isEmpty(self):
        return self.size == 0
    
    def __isSingle(self):
        return self.size == 1
    
    def __contains(self, value):
        return self.indexOf(value) >= 0
    
    
    def __get_previous(self, node):
        current = self.head
        while current.next:
            if current.next == node:
                return current
            current = current.next
            
        return None
                
    
    def indexOf(self, value):
        current = self.head
        count = 0
        while current:
            if current.value == value:
                return count
            current = current.next
            count += 1
        return -1
            
            
    def print_out(self):
        current = self.head
        while current:
            print(current.value, end = " ")
            current = current.next
        print(" ")
        
        
    def add_first(self, value):
        node = Node(value)
        node.next = self.head
        self.head = node
        self.size += 1
    
    
    def add_last(self, value):
        node = Node(value)
        if self.__isEmpty():
            self.head = node
        else: 
            current = self.head
            while current.next is not None:
                current = current.next
            current.next = node
        self.size += 1
        
        
    def delete_first(self):
        if self.__isEmpty():
            raise Exception ("List is empty!")
        else:  
            second = self.head.next
            self.head.next = None
            self.head = second
            
        
    def delete_last(self):
        if self.__isEmpty():
            raise Exception ("List is empty!")
        if self.__isSingle():
            self.head = None
        else:
            current = self.head
            while current.next:
                current = current.next
            previous = self.__get_previous(current)
            previous.next = None
            self.size -= 1
            
                
    def delete_at(self, index):
        if index < 0 or index > self.size:
            raise Exception ("Invalid index!!")
        
        count = 0        
        current = self.head
        while current:
            if count == index:
                previous = self.__get_previous(current)
                previous.next = current.next
                del current
                self.size -= 1
            else:
                current = current.next
                count += 1
       
                
    def delete_value(self, value):
        current = self.head
        while current:
            if current.value == value:
                previous = self.__get_previous(current)
                previous.next = current.next
                del current
                self.size -= 1
            else:
                current = current.next
        raise Exception ("There is no {} in the list!".format(value))
                
                
    def clear(self):
        while self.__isEmpty() == False:
            current = self.head
            self.head = self.head.next
            del current
            self.size -= 1
            
            
    def reverse(self):
        if (self.__isEmpty() or self.__isSingle()): return
    
        previous = None
        current = self.head 
        while(current): 
            forward = current.next
            current.next = previous 
            previous = current 
            current = forward
        self.head = previous
        
          
    def get_kth_from_end(self, k):
        slow = self.head
        current = self.head
        if(self.__isEmpty()): return -1
        for i in range(k-1):
            current = current.next
            if current is None:
                return -1
            
        fast = current
        while fast.next:
            fast = fast.next
            slow = slow.next
        
        return slow.value
            

File: cpp/LinkedList/LinkedList.cpp
#include <iostream>
#include "LinkedList.h"
using namespace std;

int LinkedList::getSize() {
    return _size;
}

bool LinkedList::isEmpty() {
    return head == nullptr;
}

Node* LinkedList::getPrevious(Node* node) {
    Node* current = head;
    while (current->next != nullptr) {
        if (current->next == node)
            return current;
        current = current->next;
    }
    return nullptr;
}

int LinkedList::indexOf(int value) {
    int count = 0;
    Node* current = head;
    while (current != nullptr) {
        if (current->value == value)
            return count;
        current = current->next;
    }
    return -1;
}

bool LinkedList::contains(int value) {
    return indexOf(value) > 0;
}

bool LinkedList::isSingle() {
    return head->next == nullptr;
}

void LinkedList::print() {
    if (isEmpty()) {
        cout << "List is empty.\n";
        return;
    }

    Node* current = head;
    while (current != nullptr) {
        cout << current->value << " ";
        current = current->next;
    }
    cout << endl;
}

void LinkedList::addFirst(int value) {
    Node* newNode = new Node(value);
    newNode->next = head;
    head = newNode;
    _size++;
}

void LinkedList::addLast(int value) {
    Node* newNode = new Node(value);
    if (isEmpty()) {
        head = newNode;
    }
    else {
        Node* current = head;
        while (current->next != nullptr) {
            current = current->next;
        }
        current->next = newNode;
    }
    _size++;
}

void LinkedList::addAt(int idx, int val) {
    idx -= 1;
    Node* node = new Node(val);
    Node* current = head;
    while(idx--) {
        current = current->next;
    }
    node->next = current->next;
    current->next = node;
    _size++;
}

void LinkedList::deleteFirst() {
    if (isEmpty()) {
        cout << "List is empty.\n";
        return;
    }
    if (isSingle()) {
        head = nullptr;
    }

    Node* second = head->next;
    head->next = nullptr;
    head = second;
    _size--;
}

void LinkedList::deleteLast() {
    if (isEmpty()) {
        cout << "List is empty.\n";
        return;
    }
    if (isSingle()) {
        head = nullptr;
    }
    Node* current = head;
    while (current != nullptr) {
        if (current->next == nullptr) {
            Node* previous = getPrevious(current);
            previous->next = nullptr;
        }
        current = current->next;
    }
    _size--;
}

void LinkedList::deleteAt(int index) {
    if (index >= _size || index < 0) {
        return;
    }
    Node* current = head;
    index -= 1;
    while(index--) {
        current = current ->next;
    }
    Node* toDelete = current->next;
    current->next = current->next->next;
    delete toDelete;
    _size--;
}

void LinkedList::deleteValue(int value) {
    Node* current = head;
    while (current != nullptr) {
        if (current->value == value) {
            Node* previous = getPrevious(current);
            previous->next = current->next;
            delete current;
            _size--;
        }
        current = current->next;
    }
    cout << "There is no " << value << " in list!" << endl;
}

void LinkedList::clear() {
    while (!isEmpty()) {
        Node* current = head;
        head = head->next;
        delete current;
        _size--;
    }
}

void LinkedList::reverse() {
    if (isEmpty() || isSingle())
        return;

    Node* current = head;
    Node* prev = nullptr;
    Node* next = nullptr;
    while (current) {
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    head = prev;
}

int LinkedList::getKthFromEnd(int k) {
    Node* slow = head;
    Node* current = head;
    if (isEmpty())
        return -1;
    for (int i = 0; i < k - 1; i++) {
        current = current->next;
        if (current == nullptr)
            return -1;
    }

    Node* fast = current;
    while (fast->next != nullptr) {
        fast = fast->next;
        slow = slow->next;
    }
    return slow->value;
}

Node* LinkedList::swapPairs(Node* head) {
    if(!head) return NULL;
    Node* tmp = new Node(0);
    tmp->next = head;
    Node* pre = tmp, *cur = head;
    while(cur && cur->next) {
        pre->next = cur->next;
        pre = pre->next;
        cur->next = pre->next;
        pre->next = cur;
        pre = cur;
        cur = cur->next;
    }
    return tmp->next;
}


int main() {

    LinkedList list;
    list.print();
    list.addLast(5);
    list.addLast(3);
    list.addFirst(9);
    list.print();
    list.addLast(4);
    list.print();

    cout << "From last 2: " << list.getKthFromEnd(2) << endl;
    list.deleteAt(2);
    list.print();

    list.addAt(2, 7);
    list.print();

    list.deleteFirst();
    list.print();

    list.deleteLast();
    list.print();

    list.addFirst(8);
    list.print();
    list.reverse();
    list.print();

    list.deleteValue(10);
    list.print();

    list.clear();
    list.print(); // print: List is empty.

    return 0;
}
